include/git2/mailmap.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+#ifndef INCLUDE_git_mailmap_h__
+#define INCLUDE_git_mailmap_h__
+
+#include "common.h"
+#include "types.h"
+#include "buffer.h"
+
+/**
+ * @file git2/mailmap.h
+ * @brief Mailmap parsing routines
+ * @defgroup git_mailmap Git mailmap routines
+ * @ingroup Git
+ * @{
+ */
+GIT_BEGIN_DECL
+
+/**
+ * Allocate a new mailmap object.
+ *
+ * This object is empty, so you'll have to add a mailmap file before you can do
+ * anything with it. The mailmap must be freed with 'git_mailmap_free'.
+ *
+ * @param out pointer to store the new mailmap
+ * @return 0 on success, or an error code
+ */
+GIT_EXTERN(int) git_mailmap_new(git_mailmap **out);
+
+/**
+ * Free the mailmap and its associated memory.
+ *
+ * @param mm the mailmap to free
+ */
+GIT_EXTERN(void) git_mailmap_free(git_mailmap *mm);
+
+/**
+ * Add a single entry to the given mailmap object. If the entry already exists,
+ * it will be replaced with the new entry.
+ *
+ * @param mm mailmap to add the entry to
+ * @param real_name the real name to use, or NULL
+ * @param real_email the real email to use, or NULL
+ * @param replace_name the name to replace, or NULL
+ * @param replace_email the email to replace
+ * @return 0 on success, or an error code
+ */
+GIT_EXTERN(int) git_mailmap_add_entry(
+	git_mailmap *mm, const char *real_name, const char *real_email,
+	const char *replace_name, const char *replace_email);
+
+/**
+ * Parse mailmap entries from a buffer.
+ *
+ * @param mm mailmap to add the entries to
+ * @param buf the buffer to read the mailmap file from
+ * @param len the length of the input buffer [optional: 0 defaults to 'strlen']
+ * @return 0 on success, or an error code
+ */
+GIT_EXTERN(int) git_mailmap_add_buffer(
+	git_mailmap *mm, const char *buf, size_t len);
+
+/**
+ * Create a new mailmap instance containing a single mailmap file
+ *
+ * This method is a simple utility wrapper for the following sequence
+ * of calls:
+ *  - git_mailmap_new
+ *  - git_mailmap_add_buffer
+ *
+ * @param out pointer to store the new mailmap
+ * @param buf buffer to parse the mailmap from
+ * @param len the length of the input buffer [optional: 0 defaults to 'strlen']
```

Comment:
This is not a pattern that we have been using - our APIs either take a NUL-terminated string or a buffer with length.  (And if it's the latter, then `0` really means `0`.)  I think it would be confusing to have an API that differs from this pattern.

I don't have a strong opinion on which way this should go - it seems like we should just take a `buf`?  I'm happy either way but the `0` means we do the `strlen` doesn't align with the rest of our APIs.

Reply:
It used to take a buf, and I changed it in an earlier patch. I'm fine with switching back to that if it's what we want to do.